frontend/lib/resolution/scope-queries.cpp

```diff
@@ -511,20 +511,58 @@ static bool doLookupInScope(Context* context,
     return false;
   }
 
-  if (checkDecls) {
-    bool got = scope->lookupInScope(name, result, skipPrivateVisibilities);
+  // Get the resolved visibility statements, if available
+  const ResolvedVisibilityScope* r = nullptr;
+  if (checkUseImport) {
+    if (resolving && resolving->scope() == scope) {
+      r = resolving;
+    } else {
+      r = resolveVisibilityStmts(context, scope);
+    }
+  }
+
+  // gather non-shadow scope information
+  // (declarations in this scope as well as public use / import)
+  {
+    bool got = false;
+    if (checkDecls) {
+      got |= scope->lookupInScope(name, result, skipPrivateVisibilities);
+    }
+    if (checkUseImport) {
+      got |= doLookupInImportsAndUses(context, scope, resolving, r, name,
+                                      onlyInnermost, skipPrivateVisibilities,
+                                      VisibilitySymbols::REGULAR_SCOPE,
+                                      checkedScopes, result);
+    }
     if (onlyInnermost && got) return true;
   }
 
-  // Look at use/import statements in the current scope
+  // now check shadow scope 1 (only relevant for 'private use')
   if (checkUseImport) {
     bool got = false;
-    got = doLookupInImports(context, scope, resolving,
-                            name, onlyInnermost, skipPrivateVisibilities,
-                            checkedScopes, result);
+    got |= doLookupInImportsAndUses(context, scope, resolving, r, name,
+                                    onlyInnermost, skipPrivateVisibilities,
+                                    VisibilitySymbols::SHADOW_SCOPE_ONE,
+                                    checkedScopes, result);
+
+    // treat the auto-used modules as if they were 'private use'd
+    got |= doLookupInAutoModules(context, scope, resolving, name,
+                                 onlyInnermost, skipPrivateVisibilities,
+                                 checkedScopes, result);
     if (onlyInnermost && got) return true;
   }
 
+  // now check shadow scope 2 (only relevant for 'private use')
```

Comment:
Might these three checks be somewhat elegantly combined into a for-loop for `int8_t` and `VisibilitySymbols::ShadowScope`? something like:

```c++
for (int8_t i = VisibilitySymbols::REGULAR_SCOPE; i < VisibilitySymbols::SHADOW_SCOPE_TWO; i++) {
  bool got = false;
  if (i == VisibilitySymbols::REGULAR_SCOPE) { /* lookupInScope logic */ }
  if (i == VisibilitySymbols::SHADOW_SCOPE_ONE) { /* auto-use logic */ }
  got |= doLookupInImportsAndUses(context, scope, resolving, r, name,
                                 onlyInnermost, skipPrivateVisibilities,
                                 VisibilitySymbols::SHADOW_SCOPE_TWO,
                                 checkedScopes, result);
  if (onlyInnermost && got) return true;
}
```
I'm on the fence about this suggestion, but the reason I like it is that it emphasizes the fact that we are searching each of the three shadow scopes in turn, though some have additional things to search.

Reply:
Yes, I agree that it could be written that way. However, I feel that the comments adequately emphasize that we are searching each in turn (and if you have a suggestion for improving those comments, I'm happy to do so). At the same time, having a `for` loop that has conditionals to create different behavior for every index visited seems a bit of an anti-pattern as well.

So, let's leave it as-is.